ocaml/xapi/repository_helpers.ml

```diff
@@ -1606,3 +1606,44 @@ let update_livepatch_failures ~__context ~host ~applied ~failed =
   List.iter host_remove
     (List.filter_map Guidance.to_pending_guidance to_be_removed) ;
   List.iter host_add (List.filter_map Guidance.to_pending_guidance to_be_added)
+
+let assert_no_host_pending_mandatory_guidance ~__context ~host =
+  match Db.Host.get_pending_guidances ~__context ~self:host with
+  | [] ->
+      ()
+  | _ :: _ ->
+      raise
+        Api_errors.(
+          Server_error
+            (host_pending_mandatory_guidances_not_empty, [Ref.string_of host])
+        )
+
+let assert_host_evacuation_if_required ~__context ~host ~mandatory =
+  let open Guidance in
+  let need_evacuation =
+    List.exists (fun g -> g = RebootHost || g = EvacuateHost) mandatory
+  in
+  let resident_vms =
+    Db.Host.get_resident_VMs ~__context ~self:host
+    |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
+    |> List.filter_map (fun (ref, record) ->
+           match
+             ( record.API.vM_is_control_domain
+             , record.API.vM_power_state
```

Comment:
The power-state check isn't really needed: if a VM is halted or suspended, then resident_on will always be null. You will need to excluded the control domain. Instead of getting the whole record above, you could just get `is_control_domain`.